File: mudlib/daemons/editor_d.lua
-- mudlib/daemons/editor_d.lua — Typing more than one line.
--
-- A room description is six lines of prose, and until now there was no way to
-- enter one. `pager_d` is output-only; `login.lua` is a state machine reachable
-- from `on_input` and not from a command; `game/cmds/board.lua` fakes a
-- two-field post with a `|` pipe and says so in a comment — *"a board that needs
-- an editor session to post one line is a board nobody posts to."*
--
-- `docs/src/lua-api/olc.md` argued against building one: *"An in-game editor for
-- prose would be a worse text editor than the one you already have."* That is
-- correct about a **text editor** and wrong about **a way to type six lines**.
-- The alternatives are a `|`-delimited one-liner, which becomes unusable around
-- line four, and "go and edit the file", which is what OLC exists to avoid.
--
-- ─── The shape ───────────────────────────────────────────────────────────────
--
-- Exactly `pager_d`'s: a per-session table, an `is_editing` predicate, a
-- `handle_input`, and one interception at the top of `Commands.dispatch` —
-- immediately *after* the pager's. Two interceptors in a fixed order is easier
-- to reason about than one that arbitrates between them.
--
-- Commands are dot-prefixed so a line of prose is never eaten. `quit` typed in
-- here is *text*, deliberately: a description containing the word is ordinary,
-- and losing an hour's writing to it would not be. `.q` is the only way out.

local M = {}

-- session_id → { title, lines = {}, on_save, on_abort }
local _editing = {}

--- A bound on one buffer. Prose, not a file: a description forty lines long is
--- a mistake, and a runaway paste should stop rather than grow without limit.
M.MAX_LINES = 200

local function log_error(message)
    log("error", message)
    if DAEMON and DAEMON.journal then pcall(DAEMON.journal.error, message) end
end

local HELP = {
    "{cyan}Editor{/}  — type prose; lines beginning with a dot are commands.",
    "  {yellow}.s{/} / {yellow}.save{/}    save and close",
    "  {yellow}.q{/} / {yellow}.abort{/}   discard and close",
    "  {yellow}.l{/}            list with line numbers",
    "  {yellow}.d <n>{/}        delete line n",
    "  {yellow}.i <n> <text>{/} insert before line n",
    "  {yellow}.c{/}            clear the buffer",
    "  {yellow}.h{/}            this help",
    "  {yellow}..{/}            a literal line starting with a dot",
}

--- Speak to the session, whether or not a player is behind it.
---
--- `get_player` *raises* on an id that is not a real session, so it is guarded
--- rather than merely checked. The editor is usable from a daemon and from a
--- test with a synthetic id, and a helper that only worked for a logged-in
--- player would make both impossible.
local function send_to(session_id, text)
    local ok, player = pcall(get_player, session_id)
    if ok and player then
        player:send(text)
    else
        pcall(send, session_id, text .. "\r\n")
    end
end

--- Show the buffer with line numbers.
local function list(session_id, state)
    if #state.lines == 0 then
        send_to(session_id, "{yellow}(empty){/}")
        return
    end
    local out = {}
    for i, line in ipairs(state.lines) do
        out[#out + 1] = string.format("{yellow}%3d{/}] %s", i, line)
    end
    local ok, player = pcall(get_player, session_id)
    if ok and player then
        -- `literal`, because the buffer is the builder's prose: a `{red}` they
        -- typed is text they typed, not markup for us to render.
        player:send_paged(table.concat(out, "\r\n"), { literal = true })
    else
        pcall(send, session_id, table.concat(out, "\r\n") .. "\r\n")
    end
end

--- Open an editing session.
---
--- `initial` is pre-loaded, so editing a description is editing rather than
--- retyping. That is most of the difference between an editor somebody uses and
--- one they work around.
--- @param session_id string
--- @param opts table  { title, initial = string|table, on_save, on_abort }
--- @return boolean
function M.open(session_id, opts)
    opts = opts or {}
    if _editing[session_id] then
        send_to(session_id, "{red}You are already editing something.{/}")
        return false
    end

    local lines = {}
    if type(opts.initial) == "string" and opts.initial ~= "" then
        for line in (opts.initial .. "\n"):gmatch("(.-)\r?\n") do
            lines[#lines + 1] = line
        end
        -- Drop the empty tail `gmatch` leaves on a string ending in a newline.
        if #lines > 0 and lines[#lines] == "" then lines[#lines] = nil end
    elseif type(opts.initial) == "table" then
        for _, line in ipairs(opts.initial) do lines[#lines + 1] = line end
    end

    _editing[session_id] = {
        title    = opts.title or "text",
        lines    = lines,
        on_save  = opts.on_save,
        on_abort = opts.on_abort,
    }

    send_to(session_id, "{cyan}[EDITOR]{/} " .. (opts.title or "text")
        .. " — " .. #lines .. " line" .. (#lines == 1 and "" or "s")
        .. ".  {yellow}.s{/} save  {yellow}.q{/} abort  {yellow}.h{/} help")
    if #lines > 0 then list(session_id, _editing[session_id]) end
    return true
end

--- Is this session mid-edit?
--- @param session_id string
--- @return boolean
function M.is_editing(session_id)
    return _editing[session_id] ~= nil
end

--- Close without saving, running no callback. For cleanup paths.
--- @param session_id string
function M.stop(session_id)
    _editing[session_id] = nil
end

--- Close and run `on_abort`.
--- @param session_id string
function M.abort(session_id)
    local state = _editing[session_id]
    _editing[session_id] = nil
    if state and type(state.on_abort) == "function" then
        local ok, err = pcall(state.on_abort)
        if not ok then log_error("EDITOR_D: on_abort raised: " .. tostring(err)) end
    end
end

--- Close and run `on_save` with the joined text.
--- @param session_id string
function M.save(session_id)
    local state = _editing[session_id]
    if not state then return end
    _editing[session_id] = nil

    local text = table.concat(state.lines, "\n")
    send_to(session_id, "{green}[EDITOR]{/} Saved " .. #state.lines .. " line"
        .. (#state.lines == 1 and "" or "s") .. ", " .. #text .. " characters.")

    if type(state.on_save) == "function" then
        local ok, err = pcall(state.on_save, text, state.lines)
        if not ok then
            log_error("EDITOR_D: on_save raised: " .. tostring(err))
            send_to(session_id, "{red}[EDITOR] Saving failed. See the log.{/}")
        end
    end
end

--- One line of input, while editing.
---
--- Called from `Commands.dispatch` before anything else looks at the text, so a
--- line of prose is never parsed as a verb.
--- @param session_id string
--- @param text string
function M.handle_input(session_id, text)
    local state = _editing[session_id]
    if not state then return end
    text = text or ""

    -- `..` escapes a literal leading dot, so a line of prose may begin with one.
    if text:sub(1, 2) == ".." then
        text = text:sub(2)
    elseif text:sub(1, 1) == "." then
        local cmd, rest = text:match("^%.(%a*)%s*(.*)$")
        cmd = (cmd or ""):lower()

        if cmd == "s" or cmd == "save" then
            return M.save(session_id)

        elseif cmd == "q" or cmd == "abort" then
            send_to(session_id, "{yellow}[EDITOR]{/} Discarded.")
            return M.abort(session_id)

        elseif cmd == "l" or cmd == "list" then
            list(session_id, state)
            return

        elseif cmd == "c" or cmd == "clear" then
            state.lines = {}
            send_to(session_id, "{yellow}[EDITOR]{/} Buffer cleared.")
            return

        elseif cmd == "h" or cmd == "help" then
            local ok, player = pcall(get_player, session_id)
            if ok and player then player:send_lines(HELP) end
            return

        elseif cmd == "d" then
            local n = tonumber(rest)
            if not n or not state.lines[n] then
                send_to(session_id, "{red}[EDITOR] No line " .. tostring(rest) .. ".{/}")
            else
                table.remove(state.lines, n)
                send_to(session_id, "{yellow}[EDITOR]{/} Deleted line " .. n .. ".")
            end
            return

        elseif cmd == "i" then
            local n, body = rest:match("^(%d+)%s?(.*)$")
            n = tonumber(n)
            if not n or n < 1 or n > #state.lines + 1 then
                send_to(session_id, "{red}[EDITOR] Usage: .i <line> <text>{/}")
            else
                table.insert(state.lines, n, body or "")
                send_to(session_id, "{yellow}[EDITOR]{/} Inserted before line " .. n .. ".")
            end
            return

        else
            send_to(session_id, "{red}[EDITOR] Unknown command '." .. cmd
                .. "'. `.h` for help, `..` for a literal dot.{/}")
            return
        end
    end

    if #state.lines >= M.MAX_LINES then
        send_to(session_id, "{red}[EDITOR] " .. M.MAX_LINES
            .. " lines is the limit. `.s` to save, `.d <n>` to make room.{/}")
        return
    end

    state.lines[#state.lines + 1] = text
end

--- Forget a session. Called from on_disconnect.
---
--- An editor left open by a dropped connection would wedge that session for
--- ever: every subsequent line would be buffered as prose into a buffer nobody
--- will ever save.
--- @param session_id string
function M.cleanup(session_id)
    _editing[session_id] = nil
end

log("debug", "EDITOR_D: daemon loaded")

return M

File: tests/mudlib/editor_d.rs
//! The line editor, and the one property that makes it safe.
//!
//! Nothing in the repository could accept multi-line input: `pager_d` is
//! output-only, `login.lua` is reachable only from `on_input`, and
//! `game/cmds/board.lua` fakes a two-field post with a `|` pipe because of it. A
//! room description is six lines of prose, so OLC needed one.
//!
//! What it must not do is *lose* input. While it is open every line is text, and
//! the interception has to happen above command dispatch — otherwise typing
//! "look at the far wall" in the middle of a description runs `look`.

use crate::common::RealVm;

/// While the editor is open, a verb is **text**, not a command.
///
/// The property the whole design rests on. `quit` inside a description is
/// ordinary English, and losing an hour's writing to it would not be — so `.q`
/// is the only way out, and nothing else is interpreted.
#[test]
fn a_verb_typed_into_the_editor_is_buffered_not_executed() {
    let mut vm = RealVm::boot_with_fixture_world(0);
    vm.command("pagesize 0");

    vm.lua(
        "DAEMON.editor.open(SESSION, { title = 'test', on_save = function(text) SAVED = text end })",
    );

    // Three lines that are all real commands.
    let out = vm.command("look");
    assert!(
        !out.contains("Obvious exits"),
        "`look` was executed instead of buffered:\n{out}"
    );
    vm.command("quit");
    vm.command("who");

    vm.command(".s");
    assert_eq!(
        vm.lua("return SAVED"),
        "look\nquit\nwho",
        "the buffer should hold exactly what was typed"
    );

    // …and the session is still here, which `quit` would have ended.
    assert!(vm.command("look").contains("Obvious exits"), "the session ended");
}

/// A blank line is a paragraph break, not nothing.
///
/// The interception sits *above* dispatch's empty-line check for this reason: an
/// editor that swallows blank lines cannot type a paragraph.
#[test]
fn a_blank_line_is_kept() {
    let mut vm = RealVm::boot_with_fixture_world(0);
    vm.lua(
        "DAEMON.editor.open(SESSION, { on_save = function(t) SAVED = t end })",
    );
    vm.command("First paragraph.");
    vm.command("");
    vm.command("Second paragraph.");
    vm.command(".s");

    assert_eq!(vm.lua("return SAVED"), "First paragraph.\n\nSecond paragraph.");
}

/// `..` escapes a leading dot, so prose may begin with one.
#[test]
fn a_literal_leading_dot_can_be_typed() {
    let mut vm = RealVm::boot_with_fixture_world(0);
    vm.lua("DAEMON.editor.open(SESSION, { on_save = function(t) SAVED = t end })");
    vm.command("..s is how you save.");
    vm.command(".s");

    assert_eq!(vm.lua("return SAVED"), ".s is how you save.");
}

/// The buffer is pre-loaded, so editing is editing rather than retyping.
#[test]
fn an_existing_value_is_loaded_for_editing() {
    let mut vm = RealVm::boot_with_fixture_world(0);
    vm.command("pagesize 0");

    // Straight through `command`, because what is being checked here is the
    // *banner* rather than a returned value — `lua` gives you the latter.
    let out = vm.command(
        "fixtureeval DAEMON.editor.open(SESSION, { title = 'crypt.hall.description', \
           initial = 'Line one.\\nLine two.', on_save = function(t) SAVED = t end })",
    );
    assert!(out.contains("2 lines"), "{out}");
    assert!(out.contains("Line one."), "the existing text should be shown:\n{out}");

    // Append and save: what comes back is the old text plus the new line.
    vm.command("Line three.");
    vm.command(".s");
    assert_eq!(vm.lua("return SAVED"), "Line one.\nLine two.\nLine three.");
}

/// The line commands do what they say.
#[test]
fn lines_can_be_deleted_inserted_and_cleared() {
    let mut vm = RealVm::boot_with_fixture_world(0);
    vm.command("pagesize 0");
    vm.lua(
        "DAEMON.editor.open(SESSION, { initial = 'a\\nb\\nc', on_save = function(t) SAVED = t end })",
    );

    vm.command(".d 2");
    vm.command(".i 1 zero");
    vm.command(".s");
    assert_eq!(vm.lua("return SAVED"), "zero\na\nc");

    // `.c` empties it, and `.q` discards rather than saving.
    vm.lua(
        "ABORTED = false DAEMON.editor.open(SESSION, { initial = 'x', \
           on_save = function() SAVED = 'saved' end, on_abort = function() ABORTED = true end })",
    );
    vm.command(".c");
    vm.command(".q");
    assert_eq!(vm.lua("return tostring(ABORTED)"), "true");
    assert_eq!(vm.lua("return SAVED"), "zero\na\nc", "abort must not save");
}

/// An unknown dot-command is refused rather than buffered.
///
/// Buffering it would put `.savee` in the middle of somebody's prose and look
/// like the editor had simply ignored them.
#[test]
fn an_unknown_dot_command_is_refused() {
    let mut vm = RealVm::boot_with_fixture_world(0);
    vm.lua("DAEMON.editor.open(SESSION, { on_save = function(t) SAVED = t end })");

    let out = vm.command(".savee");
    assert!(out.contains("Unknown command"), "{out}");
    vm.command(".s");
    assert_eq!(vm.lua("return SAVED"), "", "the typo should not be in the buffer");
}

/// Closing releases the session, so commands work again.
#[test]
fn closing_the_editor_gives_the_session_back() {
    let mut vm = RealVm::boot_with_fixture_world(0);

    vm.lua("DAEMON.editor.open(SESSION, {})");
    assert!(
        !vm.command("look").contains("Obvious exits"),
        "the editor should be holding input"
    );
    vm.command(".q");
    assert!(vm.command("look").contains("Obvious exits"), "the session stayed held");
}

// ─── daemon-level, through the probe VM ──────────────────────────────────────
//
// `fixtureeval` is a *command*, so while the editor is open it is buffered as
// prose like everything else — which is the feature working. Anything that has
// to inspect editor state while it is open therefore asks the daemon directly,
// where dispatch is not involved at all.

/// A dropped connection releases the editor.
///
/// Left open, every subsequent line from a reconnecting session would be
/// buffered as prose into a buffer nobody will ever save — a wedged session with
/// no visible cause and no way out.
#[test]
fn a_disconnect_releases_the_editor() {
    let mut vm = RealVm::boot_fixture_with_probe();

    assert_eq!(
        vm.eval("DAEMON.editor.open('s1', {}) return tostring(DAEMON.editor.is_editing('s1'))")
            .unwrap(),
        "true"
    );
    assert_eq!(
        vm.eval("DAEMON.editor.cleanup('s1') return tostring(DAEMON.editor.is_editing('s1'))")
            .unwrap(),
        "false"
    );

    // `cleanup` runs no callback — a dropped connection is not a save and not
    // an abort, and firing either would be inventing an intention.
    assert_eq!(
        vm.eval(
            "FIRED = false DAEMON.editor.open('s2', { on_abort = function() FIRED = true end }) \
             DAEMON.editor.cleanup('s2') return tostring(FIRED)"
        )
        .unwrap(),
        "false"
    );
}

/// The buffer is bounded, so a runaway paste stops rather than growing.
#[test]
fn the_buffer_has_a_limit() {
    let mut vm = RealVm::boot_fixture_with_probe();

    let out = vm
        .eval(
            "local n = DAEMON.editor.MAX_LINES \
             DAEMON.editor.open('s1', { on_save = function(_, lines) COUNT = #lines end }) \
             for i = 1, n + 5 do DAEMON.editor.handle_input('s1', 'line ' .. i) end \
             DAEMON.editor.save('s1') return COUNT .. '|' .. n",
        )
        .unwrap();
    let (kept, cap) = out.split_once('|').unwrap();
    assert_eq!(kept, cap, "the buffer grew past its limit");
}

/// One session at a time. Opening over an open editor would silently discard
/// whatever was already in it.
#[test]
fn a_second_open_is_refused() {
    let mut vm = RealVm::boot_fixture_with_probe();

    assert_eq!(
        vm.eval(
            "DAEMON.editor.open('s1', { initial = 'kept' }) \
             local second = DAEMON.editor.open('s1', { initial = 'lost' }) \
             DAEMON.editor.open = DAEMON.editor.open \
             local held = '' \
             DAEMON.editor.stop('s1') \
             return tostring(second)"
        )
        .unwrap(),
        "false"
    );
}
